File: cowichan/baSrcPaper/go/chain/seq/src/all/outer.go
/*
 * outer: outer product
 *
 * input:
 *   Outer_vector: a vector of (x, y) points
 *   nelts: the number of points
 *
 * output:
 *   Outer_matrix: a real matrix, whose values are filled with inter-point
 *     distances
 *   Outer_vector: a real vector, whose values are filled with origin-to-point
 *     distances
 */
package all

import (
  "math"
)

type Double float64;

var Outer_matrix [][]Double;
var Outer_vector []Double;

func max(a, b Double) Double {
  if a > b {
    return a;
  }
  return b;
}

func sqr(x Double) Double {
  return x * x;
}

func distance(a, b WinnowPoint) Double {
  return Double(math.Sqrt(float64(sqr(Double(a.i - b.i)) +
        sqr(Double(a.j - b.j)))));
}

func Outer(nelts int) {
  Outer_matrix = make ([][]Double, nelts)
  for i := range Outer_matrix {
    Outer_matrix [i] = make ([]Double, nelts)
  }

  Outer_vector = make ([]Double, nelts)

  for i := 0; i < nelts; i++ {
    var nmax Double = -1;
    for j := 0; j < nelts; j++ {
      if (i != j) {
        Outer_matrix[i][j] = distance(Winnow_points[i], Winnow_points[j]);
        nmax = max(nmax, Outer_matrix[i][j]);
      }
    }
    Outer_matrix[i][i] = Double(nelts) * nmax;
    Outer_vector[i] = distance(WinnowPoint{i: 0, j: 0}, Winnow_points[i]);
  }
}

File: cowichan/baSrcPaper/tbb/chain/par/outer.cc
/* outer: outer product
 *
 * input:
 *   winnow_points: a vector of (x, y) points
 *   nelts: the number of points
 *
 * output:
 *   outer_matrix: a real matrix, whose values are filled with inter-point
 *     distances
 *   outer_vector: a real vector, whose values are filled with origin-to-point
 *     distances
 */
#include <cassert>
#include <cmath>
#include <cstdio>
#include <cstdlib> 
#include <cstring>

#include <algorithm>
#include <iostream>
#include <vector>

#include "tbb/blocked_range.h"
#include "tbb/parallel_for.h"
#include "tbb/task_scheduler_init.h"

using namespace std;
using namespace tbb;

extern pair<int, int> *winnow_points;
double *outer_matrix;
double *outer_vector;

typedef blocked_range<size_t> range;

double sqr(double x) {
  return x * x;
}

double distance(const pair<int, int>& x, const pair<int, int>& y) {
  return sqrt(sqr(x.first - y.first) + sqr(x.second - y.second));
}

void outer(int nelts) {
  outer_matrix = (double*) malloc (sizeof(double) * nelts * nelts);
  outer_vector = (double*) malloc (sizeof(double) * nelts);

  parallel_for(
    range(0, nelts),
    [&](range r) {
      for (size_t i = r.begin(); i != r.end(); ++i) {
        double nmax = -1;
        for (int j = 0; j < nelts; j++) {
          if (i != j) {
            outer_matrix[i*nelts + j] = ::distance(winnow_points[i], winnow_points[j]);
            nmax = max(nmax, outer_matrix[i*nelts + j]);
          }
        }
        outer_matrix[i*nelts + i] = nelts * nmax;
        outer_vector[i] = ::distance(make_pair(0, 0), winnow_points[i]);
      }
    });
}
